### PR TITLE
[libc++][C++03] Use `__cxx03/` headers in C++03 mode

### PR DESCRIPTION
**Pull request by philnik777**
This patch implements the forwarding to frozen C++03 headers as discussed in https://discourse.llvm.org/t/rfc-freezing-c-03-headers-in-libc. In the RFC, we initially proposed selecting the right headers from the Clang driver, however consensus seemed to steer towards handling this in the library itself. This patch implements that direction.

At a high level, the changes basically amount to making each public header look like this:

```
// inside <vector>
#ifdef _LIBCPP_CXX03_LANG
#  include <__cxx03/vector>
#else
  // normal <vector> content
#endif
```

In most cases, public headers are simple umbrella headers so there isn't much code in the #else branch. In other cases, the #else branch contains the actual implementation of the header.

### CODE REVIEW

**Comment by philnik777** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109002?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#109002** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109002?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#109001** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109001?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109000** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109000?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#108999** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/108999?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @philnik777 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by Zingam** [comment]
@philnik777 These changes seem significant enough to affect already opened PRs and to confuse the authors who are unaware of the changes.

So basically you are surrounding all C++11 mode headers with:

```C++
#include <__configuration/cxx03.h>

#if defined(_LIBCPP_CXX03_LANG) && !defined(_LIBCPP_CXX03_USE_MAIN_HEADERS)
#  include <__cxx03/any>
#else
...
#endif // _LIBCPP_CXX03_LANG
```

And these should be added even to headers not available in C++03 mode, right? Is there anything else that needs to be observed/implemented in PRs? An headers?

**Comment by philnik777** [comment]
@Zingam currently you can just continue your PRs without any changes. I'll take up any slack. Right now I guard non-C++03 header to make the changes as mechanical as possible. Some of this will be reverted in later PRs.

File: libcxx/include/algorithm
Comment (ldionne) [inline libcxx/include/algorithm:1832]: I think `_LIBCPP_USE_CXX03_HEADERS` should be a 0-1 macro for consistency. Also, I think the code is currently the opposite of what you want, i.e. you probably want

```
#if defined(_LIBCPP_CXX03_LANG) && _LIBCPP_USE_CXX03_HEADERS
```

Finally, I'd suggest renaming the macro to `_LIBCPP_USE_FROZEN_CXX03_HEADERS` to be 100% non-ambiguous.

File: libcxx/cmake/caches/cxx03-headers.cmake
Comment (ldionne) [inline libcxx/cmake/caches/cxx03-headers.cmake:1]: I would probably name this new param `use-frozen-cxx03-headers`. That tells what the setting does a bit more explicitly.

File: libcxx/include/__config
Comment (ldionne) [inline libcxx/include/__config:17]: Could we move `_LIBCPP_CXX03_LANG` into `language.h` instead? And it might be worth making a separate patch since that's so easy to do.

So in many places the includes would become `#include <__configuration/language.h>` at the beginning of files.

File: libcxx/include/atomic
Comment (ldionne) [inline libcxx/include/atomic:593]:
```suggestion
#  include <__cxx03/atomic>
```

Perhaps other occurrences of the same mistake?

**Review by ldionne — COMMENTED** [review]
I think this makes sense, however I would like you to post this PR to the RFC, since we mentioned back then that the decision of Driver vs conditional includes would be best taken based on actual proposed changes. This will give this patch a bit more visibility and we can ensure that we have consensus on moving forward with this implementation strategy.

File: libcxx/include/algorithm
Comment (philnik777) [reply libcxx/include/algorithm:1832]: Since this is a macro we want to change from the test suite I think we have to keep the `defined()` version. Or would you rather have something like
```c++
#ifndef _LIBCPP_USE_FROZEN_CXX03_HEADERS
#  define _LIBCPP_USE_FROZEN_CXX03_HEADERS 0
#endif
```
inside a config header?

File: libcxx/include/atomic
Comment (philnik777) [reply libcxx/include/atomic:593]: I've checked the top level headers manually and fixed a few more.

File: libcxx/include/__flat_map/sorted_unique.h
Comment (ldionne) [inline libcxx/include/__flat_map/sorted_unique.h:14]: Merge conflict?

File: libcxx/include/__config
Comment (ldionne) [inline libcxx/include/__config:158]: In the current state of this patch, the "intersection" between the 03 headers and the normal headers is basically the definition of `_LIBCPP_CXX03_LANG`, but we're including all of `__configuration/language.h` for that. I think that we should strive to make that intersection empty, because anything in that intersection can cause incompatibilities and confusion.

Since `_LIBCPP_CXX03_LANG` is so simple, I would instead just check the value of `__cplusplus` directly at the top-level, and not share anything between the 03 headers and the normal headers.

Note that `__config_site` is also something that technically falls inside that intersection and that's tricky to handle, we can discuss that separately.

File: libcxx/include/any
Comment (ldionne) [inline libcxx/include/any:84]: It would be nice for this patch to be really straightforward to review, and that would be the case if not for the formatting changes. I would suggest frontloading a patch that does

```c++
#if 0
// nothing for now
#else
// existing code
#endif
```

That will cause all the formatting stuff to happen in that patch. We can then follow up with a patch that turns these `#if 0` into something real.

**Review by ldionne — COMMENTED** [review]
Let's write down the order for landing this in smaller pieces:

1. Land the `#if 0` change which addresses the formatting issues.
2. Land the CMake changes that start installing the C++03 headers.
3. Switch to `#if __cplusplus < C++11 && USE_FROZEN_HEADERS` (or whatever), and also add the CI with XFAILs. If there are too many XFAILs, the CI bit should probably be left to a separate patch.

We can treat even C++23-only headers the same for now, by including `__cxx03/__config` in them. That's the status quo and we should start with that. However, as a follow-up, we should go through all of the >= 11 headers and make them "not care" about the C++03 frozen headers. We could in addition perhaps diagnose the use of `-std=c++03` in the normal `__config` file with an error.

File: libcxx/include/algorithm
Comment (ldionne) [reply libcxx/include/algorithm]: As discussed last time, I think we should keep this a `#if defined(...)` macro since we don't want to be forced to define it to `0` when we run the test suite by default. This is going to be a temporary macro anyway.

**Review by ldionne — COMMENTED** [review]
I want to record the discussion we just had about this patch.

As we make more progress on this project, we discover additional complexity that we need to handle. This was expected, however we didn't know what we were going to find. Anything that falls in the intersection of the C++03 headers and the new headers is where things can be difficult, such as:
- The configuration of the library via CMake (embodied by `__config_site`)
- The dylib, which must be shared by both sets of headers

This patch makes it clear that handling the configuration side of things is non-trivial and can lead to very confusing situations. For example, we need to shim the `__config_site` header to account for changes from `#if defined(_LIBCPP_FOO)` to `#if _LIBCPP_FOO` in the past couple of months. IMO that's extremely confusing and not a state that we can live with for a prolonged period of time, since it'll lead to bugs. The same goes for e.g. the `__locale_base_api` refactoring where we are changing some symbol names on Windows. Those are changes we'd need to backport to the C++03 headers in order for the codebase to stay maintainable.

We discussed another approach where we instead fork individual headers as we decide that we want to drop C++03 headers. For example, create `__cxx03/vector` only when we actually want to fork its implementation to take advantage of non-03 features. Seeing all the complexity introduced by the approach pursued here, that's the approach I would like for us to try instead. @philnik777 disagrees and thinks we'll end up in a similar place at the end of the day (I'm not sure I agree but that's a reasonable opinion to have).

Since the approach taken in this patch is still something we can back out of, I am okay with moving forward with it experimentally despite having significant concerns. If we don't manage to simplify the intersection between the old and the new headers, I don't think this is something we can ever ship. But if we do, then perhaps this approach is fine in the end. Since the only way to know for sure is to keep making progress on this approach, I'm fine with moving forward with this.

The subsequent patches should aim to simplify the old/new headers integration and those can be iterated upon in a separate branch without introducing too much risk for merge conflicts.

I'd like to see this again with comments resolved before this gets merged.

I feel a bit bad for being so critical of this approach, but I think it was unavoidable that we'd find out difficulties with any approach trying to do this non-trivial change. We'll have to figure out whether these difficulties can be overcome or if they actually outweigh the benefits provided by the split.

File: libcxx/include/CMakeLists.txt
Comment (ldionne) [inline libcxx/include/CMakeLists.txt:2084]:
```suggestion

```

File: libcxx/include/CMakeLists.txt
Comment (ldionne) [inline libcxx/include/CMakeLists.txt:2085]:
```suggestion
foreach(f ${files})
```

File: libcxx/include/CMakeLists.txt
Comment (ldionne) [inline libcxx/include/CMakeLists.txt:2115]:
```suggestion
  foreach(file ${files})
```

File: libcxx/include/__cxx03/__configuration/config_site_shim.h
Comment (ldionne) [inline libcxx/include/__cxx03/__configuration/config_site_shim.h:9]:
```suggestion
#ifndef _LIBCPP___CXX03___CONFIGURATION_CONFIG_SITE_SHIM_H
```

File: libcxx/utils/generate_feature_test_macro_components.py
Comment (ldionne) [inline libcxx/utils/generate_feature_test_macro_components.py:2248]: That is a stray `#endif`.

```suggestion

```

File: libcxx/test/libcxx/feature_test_macro/version_header.sh.py
Comment (ldionne) [inline libcxx/test/libcxx/feature_test_macro/version_header.sh.py:1]: I think the changes in this test file should be reverted.

File: libcxx/include/version
Comment (ldionne) [inline libcxx/include/version:1]: The header guard for `__cxx03/version` must be updated.

**Review by ldionne — APPROVED** [review]
I'm going to approve this with changes requested, and assuming CI is green. This should unblock you to keep moving forward over the holidays. I'd really like us to iterate in a separate branch until we gain a bit more confidence.

**Comment by perry-ca** [comment]
@philnik777 I noticed a few of the changes from `#include <foo>` to `#include <__cxx03/foo>` are now including files that don't exist.  I did a quick check with grep and found these headers are being included that don't exist.

- __cxx03/android/api-level.h
- __cxx03/__external_threading
- __cxx03/stdarg.h
- __cxx03/threads.h
- __cxx03/time.h
- __cxx03/unistd.h

I realize this is still a work in progress.  I hope this information helps.

**Comment by aheejin** [comment]
Hello. What’s the current status of the main headers? Has any of the C++03 support been removed from the main headers, or we can still use them in the old way by not defining `_LIBCPP_USE_FROZEN_CXX03_HEADERS`?

If it has not been removed, what’s the planned timeline for that?

**Pull request merged**
